### PR TITLE
UXD-468 Add collapsible card component

### PR DESCRIPTION
**Pull request by allison-c**
### Purpose 🚀

view examples in http://storybooks.highbond-s3.com/paprika/UXD-468-collapsible-card-component/?path=/story/navigation-collapsiblecard--variations

### Notes ✏️
# @paprika/collapsible-card - 0.0.1

## Description

Styled collapsible card component.

## Installation

```
yarn add @paprika/collapsible-card
```

or with npm:

```
npm install @paprika/collapsible-card
```

### CollapsibleCard

| Prop          | Type | required | default | Description                              |
| ------------- | ---- | -------- | ------- | ---------------------------------------- |
| children      | node | false    | null    |                                          |
| hasDivider | bool | false    | false   | If has a divider between collapsible header and content. |
| label         | node | false    | null    | Label text as the card title.            |
| onAfterExpand | func | false    | null    | Callback function after expand the card. |

### CollapsibleCard.Metadata

| Prop     | Type | required | default | Description |
| -------- | ---- | -------- | ------- | ----------- |
| children | node | false    | null    |             |

## Links

- [Storybook showcase](https://paprika.highbond.com/?path=/story/collapsible-card--showcase)
- [Github source code](https://github.com/acl-services/paprika/tree/master/packages/CollapsibleCard/src)
- [Github create issue](https://github.com/acl-services/paprika/issues/new?label=[]&title=@paprika/collapsible-card%20[help]:%20your%20short%20description&body=%0A%23%20Help%20wanted%0A%0A%23%23%20Please%20write%20your%20question.%0A*A%20clear%20and%20concise%20description%20of%20what%20the%20question%20is*%0A%0A%23%23%20Additional%20context%0A*Add%20any%20other%20context%20or%20screenshots%20about%20your%20question%20here.*%0A)
- [ChangeLog](https://github.com/acl-services/paprika/tree/master/packages/CollapsibleCard/CHANGELOG.md)

### Basic

Basic collapsible card with text label only.

```jsx
<CollapsibleCard label="Collapsible label">
  Content
</CollapsibleCard>
```

#### Result

image

### With avatar and metadata

Collapsible card with avatar (can be text or icon) and metadata

```jsx
<CollapsibleCard label="Collapsible label">
  <CollapsibleCard.Avatar>Card header</CollapsibleCard.Avatar>
  <CollapsibleCard.Metadata>1000 records</CollapsibleCard.Metadata>
  Content
</CollapsibleCard>
```

#### Result

image

### Updates 📦
- [x] MAJOR (breaking) change to `@paprika/collapsible-card`
- [ ] MINOR (backward compatible) change to _these packages_
- [x] PATCH (bug fix) change to `@paprika/collapsible`

### Storybook 📕
http://storybooks.highbond-s3.com/paprika/your-branch-name

### Screenshots 📸
_optional but highly recommended_

### References 🔗
_relevant Jira ticket / GitHub issues_

### CODE REVIEW

File: packages/CollapsibleCard/src/CollapsibleCard.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+const propTypes = {
+  children: PropTypes.node,
+
+  /** Label text as the card title. */
+  label: PropTypes.node,
+
+  /** Callback function after expand the card. */
+  onAfterExpand: PropTypes.func,
```
Comment (oscarkwan) [inline packages/CollapsibleCard/src/CollapsibleCard.js:63]: i think `onExpand` is good enough?

Comment (allison-c) [reply packages/CollapsibleCard/src/CollapsibleCard.js:63]: No, `onExpand` will be enough

Comment (allison-c) [reply packages/CollapsibleCard/src/CollapsibleCard.js:63]: I think this is mostly for the case of fetching data only if requested

Comment (oscarkwan) [reply packages/CollapsibleCard/src/CollapsibleCard.js:63]: I see, i prefer just `onExpand` for the naming then

**Comment by oscarkwan** [comment]
A11y
- not reading out collapsible text
- Not sure how i can make it so it reads out the metadata, seems like i'm not able to using screen reader

Screen Shot 2020-11-02 at 2 45 20 PM

File: packages/CollapsibleCard/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+### CollapsibleCard
+
+| Prop          | Type | required | default | Description                              |
+| ------------- | ---- | -------- | ------- | ---------------------------------------- |
+| children      | node | false    | null    |                                          |
+| label         | node | false    | null    | Label text as the card title.            |
+| onAfterExpand | func | false    | null    | Callback function after expand the card. |
+
+### CollapsibleCard.Metadata
```
Comment (oscarkwan) [inline packages/CollapsibleCard/README.md:31]: Missing `<CollapsibleCard.Avatar>`

**Comment by nahumzs** [comment]
```js
<CollapsibleCard label="Collapsible label">
  <CollapsibleCard.Avatar>Card header</CollapsibleCard.Avatar>
  <CollapsibleCard.Metadata>1000 records</CollapsibleCard.Metadata>
  <CollapsibleCard.Content>Content</CollapsibleCard.Content>
</CollapsibleCard>
```

^ that API is really prescriptive, I wonder if we should allow the user to do something like this:

```js
<CollapsibleCard label="Collapsible label">
  <MyCustomComponent />
</CollapsibleCard>
```

I feel this component will be consumed a lot and eventually Designers might want to add variations. 

For example in [these lines](https://github.com/acl-services/paprika/pull/776/files#diff-c7d011d95476d04c447c637532320e628aef83cdf90ccf418876e9c767406b0dR13-R18) you need to use `extractedChildren` to maintain the desire structure but you could do the same if you have an API like:

```js
<CollapsibleCard label="Collapsible label">
  <CollapsibleCard.Content
   label="Collapsible label" 
   avatar={<Avatar />} 
   Metadata={"1000 records"}
  >
     <Content />
  </CollapsibleCard.Content>
</CollapsibleCard>
```

The previous also allowed you to do:

```jsx
<CollapsibleCard label="Collapsible label">
  <MyOwnImplementation />
</CollapsibleCard>
```

What do you think?

**Comment by allison-c** [comment]
> A11y
> 
> * not reading out collapsible text
> * Not sure how i can make it so it reads out the metadata, seems like i'm not able to using screen reader
> 
> Screen Shot 2020-11-02 at 2 45 20 PM

This should be able to solve by passing `a11yText` through `<CollapsibleCard />` component, cause I'll pass moreProps into `<Collapsible />`. I should update at least the showcase example to show the ability.

or did you mean we should handle these automatically?

**Comment by oscarkwan** [comment]
> > A11y
> > 
> > * not reading out collapsible text
> > * Not sure how i can make it so it reads out the metadata, seems like i'm not able to using screen reader
> > 
> > Screen Shot 2020-11-02 at 2 45 20 PM
> 
> This should be able to solve by passing `a11yText` through `<CollapsibleCard />` component, cause I'll pass moreProps into `<Collapsible />`. I should update at least the showcase example to show the ability.
> 
> or did you mean we should handle these automatically?

- Even if we don't pass `a11yText` shouldn't `collapsible.a11yText` be reading actual string?

**Comment by allison-c** [comment]
> ```js
> <CollapsibleCard label="Collapsible label">
>   <CollapsibleCard.Avatar>Card header</CollapsibleCard.Avatar>
>   <CollapsibleCard.Metadata>1000 records</CollapsibleCard.Metadata>
>   <CollapsibleCard.Content>Content</CollapsibleCard.Content>
> </CollapsibleCard>
> ```
> 
> ^ that API is really prescriptive, I wonder if we should allow the user to do something like this:
> 
> ```js
> <CollapsibleCard>
>   <MyCustomComponent />
> </CollapsibleCard>
> ```
> 
> I feel this component will be consumed a lot and eventually Designers might want to add variations.
> 
> For example in [these lines](https://github.com/acl-services/paprika/pull/776/files#diff-c7d011d95476d04c447c637532320e628aef83cdf90ccf418876e9c767406b0dR13-R18) you need to use `extractedChildren` to maintain the desire structure but you could do the same if you have an API like:
> 
> ```js
> <CollapsibleCard>
>   <CollapsibleCard.Content
>    label="Collapsible label" 
>    avatar={<Avatar />} 
>    Metadata={"1000 records"}
>   >
>      <Content />
>   </CollapsibleCard.Content>
> </CollapsibleCard>
> ```
> 
> The previous also allowed you to do:
> 
> ```js-jsx
> <CollapsibleCard onExpand={handleExpand}>
>   <MyOwnImplementation />
> </CollapsibleCard>
> ```
> 
> My suggestion remove `extractedChildren` and use children for this component :)
> 
> What do you think?

I actually started with that, and then decided to go with the current API. 

> > > A11y
> > > 
> > > * not reading out collapsible text
> > > * Not sure how i can make it so it reads out the metadata, seems like i'm not able to using screen reader
> > > 
> > > Screen Shot 2020-11-02 at 2 45 20 PM
> > 
> > 
> > This should be able to solve by passing `a11yText` through `<CollapsibleCard />` component, cause I'll pass moreProps into `<Collapsible />`. I should update at least the showcase example to show the ability.
> > or did you mean we should handle these automatically?
> 
> * Even if we don't pass `a11yText` shouldn't `collapsible.a11yText` be reading actual string?

Oh I see, I'll add <L10n /> in the storybook examples

**Comment by allison-c** [comment]
> ```js
> <CollapsibleCard label="Collapsible label">
>   <CollapsibleCard.Avatar>Card header</CollapsibleCard.Avatar>
>   <CollapsibleCard.Metadata>1000 records</CollapsibleCard.Metadata>
>   <CollapsibleCard.Content>Content</CollapsibleCard.Content>
> </CollapsibleCard>
> ```
> 
> ^ that API is really prescriptive, I wonder if we should allow the user to do something like this:
> 
> ```js
> <CollapsibleCard>
>   <MyCustomComponent />
> </CollapsibleCard>
> ```
> 
> I feel this component will be consumed a lot and eventually Designers might want to add variations.
> 
> For example in [these lines](https://github.com/acl-services/paprika/pull/776/files#diff-c7d011d95476d04c447c637532320e628aef83cdf90ccf418876e9c767406b0dR13-R18) you need to use `extractedChildren` to maintain the desire structure but you could do the same if you have an API like:
> 
> ```js
> <CollapsibleCard>
>   <CollapsibleCard.Content
>    label="Collapsible label" 
>    avatar={<Avatar />} 
>    Metadata={"1000 records"}
>   >
>      <Content />
>   </CollapsibleCard.Content>
> </CollapsibleCard>
> ```
> 
> The previous also allowed you to do:
> 
> ```js-jsx
> <CollapsibleCard onExpand={handleExpand}>
>   <MyOwnImplementation />
> </CollapsibleCard>
> ```
> 
> My suggestion remove `extractedChildren` and use children for this component :)
> 
> What do you think?

I actually started with that, and then decided to go with the current API. I think I can still achieve what you want without remove <Metadata> and <Avatar> subcomponent. I decided to go with it because there're lots of components in paprika are following that and I like the clear/readable structure rather than putting everything in the props. I agree we should render all the children pass in like 
```js
 <CollapsibleCard label="Collapsible label">
   <CollapsibleCard.Avatar>Card header</CollapsibleCard.Avatar>
   <CollapsibleCard.Metadata>1000 records</CollapsibleCard.Metadata>
   Content
  More content
  <MyCustomComponent />
 </CollapsibleCard>
```
Let me try refactor them

**Comment by nahumzs** [comment]
> I actually started with that, and then decided to go with the current API. I think I can still achieve what you want without remove and subcomponent. I decided to go with it because there're lots of components in paprika are following that and I like the clear/readable structure rather than putting everything in the props. I agree we should render all the children pass in like

I think passing down children might be enough then. 👍

**Review by oscarkwan — APPROVED** [review]
LGTM

**Pull request merged**